### DISCUSSION TITLE
MappingException pops up during debugging

### DISCUSSION THREAD

**Discussion by duzenko** [doctrine/doctrine2#10019] [Support Questions]
Is there any way to not throw MappingException's internally when passing a DateTime as a parameter?
When I'm debugging my code, and a query with a date parameter executes, it throws internal exceptions and stops on that.
This is quite frustrating for me.

FWIW the exact exception message is - "Doctrine\ORM\Mapping\MappingException: Class "DateTime" is not a valid entity or mapped super class"

**Comment by greg0ire** [comment]
I know you're frustrated and all, but unless your provide a [stack trace](https://symfony.com/doc/current/contributing/code/stack_trace.html), there are little chances you're going to get any help on this. The best is to keep your feelings out of this and to stick to facts.

**Reply by duzenko** [reply]
> I know you're frustrated and all, but unless your provide a [stack trace](https://symfony.com/doc/current/contributing/code/stack_trace.html), there are little chances you're going to get any help on this. The best is to keep your feelings out of this and to stick to facts.

Sure, here's one

```
Doctrine\ORM\Mapping\Driver\AnnotationDriver->loadMetadataForClass (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/Driver/AnnotationDriver.php:102)
Doctrine\ORM\Mapping\ClassMetadataFactory->doLoadMetadata (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php:149)
Doctrine\Persistence\Mapping\AbstractClassMetadataFactory->loadMetadata (/home/a/Documents/Remote/cashare-15/vendor/doctrine/persistence/src/Persistence/Mapping/AbstractClassMetadataFactory.php:397)
Doctrine\ORM\Mapping\ClassMetadataFactory->loadMetadata (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php:79)
Doctrine\Persistence\Mapping\AbstractClassMetadataFactory->getMetadataFor (/home/a/Documents/Remote/cashare-15/vendor/doctrine/persistence/src/Persistence/Mapping/AbstractClassMetadataFactory.php:266)
Doctrine\ORM\EntityManager->getClassMetadata (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:293)
Doctrine\ORM\UnitOfWork->getSingleIdentifierValue (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:3066)
Doctrine\ORM\AbstractQuery->processParameterValue (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:435)
Doctrine\ORM\Query->resolveParameterValue (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:453)
Doctrine\ORM\Query->processParameterMappings (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:394)
Doctrine\ORM\Query->_doExecute (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:316)
Doctrine\ORM\AbstractQuery->executeIgnoreQueryCache (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:1073)
Doctrine\ORM\AbstractQuery->execute (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:1027)
Doctrine\ORM\AbstractQuery->getResult (/home/a/Documents/Remote/cashare-15/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:791)
E5\EntityRepository\BookingRepository::findUserBookings (/home/a/Documents/Remote/cashare-15/E5/EntityRepository/BookingRepository.php:107)
...
```

Query building code:
```
	static public function findUserBookings( int $userId, DateTime $fromValuta, DateTime $toValuta, string $sort = 'ASC' ): array {
		$repo = self::getInstance();
		$qb = $repo->getEntityManager()->createQueryBuilder();
		$condition = $qb->expr()->andX(
				$qb->expr()->orX(
						$qb->expr()->eq( 'b.sourceAccount', $userId ),
						$qb->expr()->eq( 'b.targetAccount', $userId ),
				),
				$qb->expr()->between( 'b.valuta', ':fromValuta', ':toValuta' ),
				$qb->expr()->orX(
						$qb->expr()->notIn( 'b.type', [
							Booking::TYPE_carryForwardUncommited,
							Booking::TYPE_payoffUncommited,
						] ),
						$qb->expr()->isNotNull( 'b.type' )
				),
		);
		$qb->select( 'b' )
				->from( $repo->getEntityName(), 'b' )
				->where( $condition )
				->addOrderBy( 'b.valuta ', $sort )
				->addOrderBy( 'b.id ', $sort )
				->setParameter( ':fromValuta', $fromValuta )
				->setParameter( ':toValuta', $toValuta );
		return $qb->getQuery()->getResult();
	}
```

The exception is thrown for each DateTime parameter, then caught and suppressed in processParameterValue()

**Reply by greg0ire** [reply]
> The exception is thrown for each DateTime parameter, then caught and suppressed in processParameterValue()

If you know where this happens, why don't you also post a permalink to that?

**Reply by duzenko** [reply]
Sorry, permalink to what exactly?

**Reply by greg0ire** [reply]
The part of the code where the catch and suppression happens, like this (I think that's what you're talking about, right? let us know): https://github.com/doctrine/orm/blob/b2a4fac40b74276dfc13fea0734d45b8fec89e98/lib/Doctrine/ORM/AbstractQuery.php#L432-L447

**Reply by duzenko** [reply]
Well, yes

**Reply by greg0ire** [reply]
> it throws internal exceptions and stops on that.

Ok so now I understand your issue better. I suppose you know that stopping on exception is something you can turn off, but that you still need it?

**Reply by duzenko** [reply]
That's the purpose of integrated debugger, right? To quickly find the code that throws because of a bug.
I mean, I understand that e.g. Java throws a lot, but Java debuggers are used to that and by default don't stop on handled exceptions - which is bad enough but still tolerable. Why would a PHP library throw, knowing that the only PHP debugger does not offer to filter caught exceptions - is a mystery to me.

**Reply by derrabus** [reply]
As long as this exception does not bubble up, this is a non-issue for me. We don't optimize our code for debuggers.

**Reply by greg0ire** [reply]
> That's the purpose of integrated debugger, right?

According to you there is no point in running a debugger when there is no exception. That's plain wrong. I would agree that using exceptions for control flow is not great, or at least unusual, and that there should be a better way to do this, but I don't see a solution. There is a method called `hasMetadataFor` that I thought could help, but the comment on it and the code shows that using it will in fact cause bugs: https://github.com/doctrine/persistence/blob/e6a1f4991cb0ad59dd332f7c3ebb93915711d0b4/src/Persistence/Mapping/ClassMetadataFactory.php#L31-L38

As @derrabus said, we don't optimize our code for debuggers so it's unlikely this is going to change.

**Reply by duzenko** [reply]
Ok, what about skipping metadata lookup for basic types like DateTime?

**Reply by greg0ire** [reply]
I think it would be reasonable to do so for any internal class, which can be detected like this: https://stackoverflow.com/a/16961772/353612 :thinking: 
Although it might be possible to define a mapping for an internal class with a non-colocated format such as XML :sweat:
We'd have to document that it's not supported.
Feel free to open a PR, but keep it mind it's likely to be closed for the reason stated above.

**Reply by duzenko** [reply]
So, is it possible or not to define a mapping for DateTime so that it does not throw in runtime? If so, how?

**Reply by greg0ire** [reply]
I don't think it is currently possible.
